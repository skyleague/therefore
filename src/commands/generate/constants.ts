import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
