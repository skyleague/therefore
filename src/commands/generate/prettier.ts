import type { ThereforeOutputType } from './types.js'

import { Nothing, isNothing } from '@skyleague/axioms'

export async function maybeLoadPrettier() {
    try {
        return await import('prettier')
    } catch {
        return Nothing
    }
}

export async function formatFile(
    prettier: Awaited<ReturnType<typeof maybeLoadPrettier>>,
    input: string,
    file: string,
    type: ThereforeOutputType
): Promise<string> {
    if (isNothing(prettier)) {
        return input
    }
    const options = await prettier.resolveConfig(file)
    return prettier.format(input, {
        ...(options ?? {}),
        ...(type === 'jsonschema' ? { parser: 'json' } : {}),
    })
}
