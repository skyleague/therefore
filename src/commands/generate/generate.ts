import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
