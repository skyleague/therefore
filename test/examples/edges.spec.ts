import { Edges } from '../../examples/edges/client.client.js'
import { compileOutput } from '../../src/commands/generate/generate.js'

import { eitherToError } from '@skyleague/axioms'
import nock from 'nock'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'

describe('client', () => {
    beforeEach(() => nock.disableNetConnect())
    afterEach(() => {
        nock.cleanAll()
        nock.enableNetConnect()
    })

    it('client', async () => {
        expect(
            await compileOutput(['examples/edges/client.schema.ts'], {
                cwd: process.cwd(),
            }),
        ).toMatchSnapshot()
    })

    it('binary responses', async () => {
        nock('http://localhost:80').get('/image').reply(200, 'foobar')

        const client = new Edges({ prefixUrl: 'http://localhost' })
        const image: unknown = eitherToError(await client.getImage())

        expect(image).toMatchInlineSnapshot(`"foobar"`)
    })

    it('yaml responses', async () => {
        nock('http://localhost:80').get('/employees').reply(200, 'foo: "bar"')

        const client = new Edges({ prefixUrl: 'http://localhost' })
        const employees: unknown = eitherToError(await client.getEmployees())

        expect(employees).toMatchInlineSnapshot(`"foo: "bar""`)
    })
})

it('locals', async () => {
    expect(
        await compileOutput(['examples/edges/locals.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})

it('optional', async () => {
    expect(
        await compileOutput(['examples/edges/optional.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})

it('validator', async () => {
    expect(
        await compileOutput(['examples/edges/validator.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})
