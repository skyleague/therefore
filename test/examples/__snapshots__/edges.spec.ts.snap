// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`client > client 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/edges/client.client.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import type { IncomingHttpHeaders } from 'node:http'

import type { DefinedError } from 'ajv'
import { got } from 'got'
import type { CancelableRequest, Got, Options, OptionsInit, Response } from 'got'

import { GetEmployeesResponse200, GetEmployeesResponseDefault } from './client.type.js'

/**
 * openapi
 */
export class DefaultTestClient {
    public client: Got

    public constructor(
    {
        prefixUrl,
        options,
        client = got
    }: {
        prefixUrl: string,
        options?: Options | OptionsInit,
        client?: Got
    }) {
        this.client = client.extend(...[{ prefixUrl, throwHttpErrors: false }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /employees
     */
    public getEmployees(): Promise<
    | SuccessResponse<"201" | "202" | "203" | "204" | "205" | "206" | "207" | "208" | "226", GetEmployeesResponseDefault>
    | SuccessResponse<'200', GetEmployeesResponse200>
    | FailureResponse<StatusCode<2>, string, 'response:body', IncomingHttpHeaders>
    | FailureResponse<StatusCode<1 | 3 | 4 | 5>, string, 'response:statuscode', IncomingHttpHeaders>
    >
    {
        return this.awaitResponse(this.client.get('employees', {
            responseType: 'json',
        }), {
            200: GetEmployeesResponse200,default: GetEmployeesResponseDefault,
        }) as ReturnType<this["getEmployees"]>
    }

    public async awaitResponse<I, S extends Record<PropertyKey, { parse: (o: I) => { left: DefinedError[] } | { right: unknown } }>>(response: 
    CancelableRequest<NoInfer<Response<I>>>
    , schemas: S) {
        const result = await response
        const status = result.statusCode < 200 ? 'informational' : result.statusCode < 300 ? 'success' : result.statusCode < 400 ? 'redirection' : result.statusCode < 500 ? 'client-error' : 'server-error'
        const validator = schemas[result.statusCode] ?? schemas.default
        const body = validator?.parse?.(result.body)
        if (result.statusCode < 200 || result.statusCode >= 300)
        {
            return {success: false as const, statusCode: result.statusCode.toString(), status, headers: result.headers, left: body !== undefined && 'right' in body ? body.right : result.body, validationErrors: body !== undefined && 'left' in body ? body.left : undefined, where: 'response:statuscode' } 
        }
        if (body === undefined || 'left' in body)
        {
            return {success: body === undefined, statusCode: result.statusCode.toString(), status, headers: result.headers, left: result.body, validationErrors: body?.left, where: 'response:body' }
        }
        return {success: true as const, statusCode: result.statusCode.toString(), status, headers: result.headers, right: result.body }
    }
}

/**
 * openapi
 */
export class DoubleSuccessClient {
    public client: Got

    public constructor(
    {
        prefixUrl,
        options,
        client = got
    }: {
        prefixUrl: string,
        options?: Options | OptionsInit,
        client?: Got
    }) {
        this.client = client.extend(...[{ prefixUrl, throwHttpErrors: false }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /employees
     */
    public getEmployees(): Promise<
    | SuccessResponse<'200', unknown>
    | FailureResponse<StatusCode<2>, string, 'response:body', IncomingHttpHeaders>
    | FailureResponse<StatusCode<1 | 3 | 4 | 5>, string, 'response:statuscode', IncomingHttpHeaders>
    >
    {
        return this.awaitResponse(this.client.get('employees', {
            responseType: 'text',
        }), {
            200: { parse: (x: unknown) => ({ right: x }) },
        }) as ReturnType<this["getEmployees"]>
    }

    /**
     * GET /image
     */
    public getImage(): Promise<
    | SuccessResponse<'200', unknown>
    | SuccessResponse<'202', unknown>
    | FailureResponse<StatusCode<2>, string, 'response:body', IncomingHttpHeaders>
    | FailureResponse<StatusCode<1 | 3 | 4 | 5>, string, 'response:statuscode', IncomingHttpHeaders>
    >
    {
        return this.awaitResponse(this.client.get('image', {
            responseType: 'text',
        }), {
            200: { parse: (x: unknown) => ({ right: x }) },202: { parse: (x: unknown) => ({ right: x }) },
        }) as ReturnType<this["getImage"]>
    }

    public async awaitResponse<I, S extends Record<PropertyKey, { parse: (o: I) => { left: DefinedError[] } | { right: unknown } }>>(response: 
    CancelableRequest<NoInfer<Response<I>>>
    , schemas: S) {
        const result = await response
        const status = result.statusCode < 200 ? 'informational' : result.statusCode < 300 ? 'success' : result.statusCode < 400 ? 'redirection' : result.statusCode < 500 ? 'client-error' : 'server-error'
        const validator = schemas[result.statusCode] ?? schemas.default
        const body = validator?.parse?.(result.body)
        if (result.statusCode < 200 || result.statusCode >= 300)
        {
            return {success: false as const, statusCode: result.statusCode.toString(), status, headers: result.headers, left: body !== undefined && 'right' in body ? body.right : result.body, validationErrors: body !== undefined && 'left' in body ? body.left : undefined, where: 'response:statuscode' } 
        }
        if (body === undefined || 'left' in body)
        {
            return {success: body === undefined, statusCode: result.statusCode.toString(), status, headers: result.headers, left: result.body, validationErrors: body?.left, where: 'response:body' }
        }
        return {success: true as const, statusCode: result.statusCode.toString(), status, headers: result.headers, right: result.body }
    }
}

/**
 * openapi
 */
export class Edges {
    public client: Got

    public constructor(
    {
        prefixUrl,
        options,
        client = got
    }: {
        prefixUrl: string,
        options?: Options | OptionsInit,
        client?: Got
    }) {
        this.client = client.extend(...[{ prefixUrl, throwHttpErrors: false }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /employees
     */
    public getEmployees(): Promise<
    | SuccessResponse<'200', unknown>
    | FailureResponse<StatusCode<2>, string, 'response:body', IncomingHttpHeaders>
    | FailureResponse<StatusCode<1 | 3 | 4 | 5>, string, 'response:statuscode', IncomingHttpHeaders>
    >
    {
        return this.awaitResponse(this.client.get('employees', {
            responseType: 'text',
        }), {
            200: { parse: (x: unknown) => ({ right: x }) },
        }) as ReturnType<this["getEmployees"]>
    }

    /**
     * GET /image
     */
    public getImage(): Promise<
    | SuccessResponse<'200', unknown>
    | FailureResponse<StatusCode<2>, string, 'response:body', IncomingHttpHeaders>
    | FailureResponse<StatusCode<1 | 3 | 4 | 5>, string, 'response:statuscode', IncomingHttpHeaders>
    >
    {
        return this.awaitResponse(this.client.get('image', {
            responseType: 'text',
        }), {
            200: { parse: (x: unknown) => ({ right: x }) },
        }) as ReturnType<this["getImage"]>
    }

    public async awaitResponse<I, S extends Record<PropertyKey, { parse: (o: I) => { left: DefinedError[] } | { right: unknown } }>>(response: 
    CancelableRequest<NoInfer<Response<I>>>
    , schemas: S) {
        const result = await response
        const status = result.statusCode < 200 ? 'informational' : result.statusCode < 300 ? 'success' : result.statusCode < 400 ? 'redirection' : result.statusCode < 500 ? 'client-error' : 'server-error'
        const validator = schemas[result.statusCode] ?? schemas.default
        const body = validator?.parse?.(result.body)
        if (result.statusCode < 200 || result.statusCode >= 300)
        {
            return {success: false as const, statusCode: result.statusCode.toString(), status, headers: result.headers, left: body !== undefined && 'right' in body ? body.right : result.body, validationErrors: body !== undefined && 'left' in body ? body.left : undefined, where: 'response:statuscode' } 
        }
        if (body === undefined || 'left' in body)
        {
            return {success: body === undefined, statusCode: result.statusCode.toString(), status, headers: result.headers, left: result.body, validationErrors: body?.left, where: 'response:body' }
        }
        return {success: true as const, statusCode: result.statusCode.toString(), status, headers: result.headers, right: result.body }
    }
}

export type Status<Major> = Major extends string ? Major extends \`1\${number}\`? 'informational': Major extends \`2\${number}\` ? 'success' : Major extends \`3\${number}\` ? 'redirection' : Major extends \`4\${number}\` ? 'client-error' : 'server-error' : undefined
export interface SuccessResponse<StatusCode extends string, T> { success: true; statusCode: StatusCode; status: Status<StatusCode>; headers: IncomingHttpHeaders; right: T }
export interface FailureResponse<StatusCode = string, T = unknown, Where = never, Headers = IncomingHttpHeaders> {
success: false
statusCode: StatusCode
status: Status<StatusCode>
headers: Headers
validationErrors: DefinedError[] | undefined
left: T
where: Where
}
export type StatusCode<Major extends number = 1 | 2 | 3 | 4 | 5> = \`\${Major}\${number}\`



/**
 * swagger
 */
export class SwaggerEdges {
    public client: Got

    public constructor(
    {
        prefixUrl,
        options,
        client = got
    }: {
        prefixUrl: string,
        options?: Options | OptionsInit,
        client?: Got
    }) {
        this.client = client.extend(...[{ prefixUrl, throwHttpErrors: false }, options].filter((o): o is Options => o !== undefined))
    }


    /**
     * GET /json-response
     */
    public getJsonResponse(): Promise<
    | SuccessResponse<'200', unknown>
    | FailureResponse<StatusCode<2>, string, 'response:body', IncomingHttpHeaders>
    | FailureResponse<StatusCode<1 | 3 | 4 | 5>, string, 'response:statuscode', IncomingHttpHeaders>
    >
    {
        return this.awaitResponse(this.client.get('json-response', {
            responseType: 'text',
        }), {
            200: { parse: (x: unknown) => ({ right: x }) },
        }) as ReturnType<this["getJsonResponse"]>
    }

    public async awaitResponse<I, S extends Record<PropertyKey, { parse: (o: I) => { left: DefinedError[] } | { right: unknown } }>>(response: 
    CancelableRequest<NoInfer<Response<I>>>
    , schemas: S) {
        const result = await response
        const status = result.statusCode < 200 ? 'informational' : result.statusCode < 300 ? 'success' : result.statusCode < 400 ? 'redirection' : result.statusCode < 500 ? 'client-error' : 'server-error'
        const validator = schemas[result.statusCode] ?? schemas.default
        const body = validator?.parse?.(result.body)
        if (result.statusCode < 200 || result.statusCode >= 300)
        {
            return {success: false as const, statusCode: result.statusCode.toString(), status, headers: result.headers, left: body !== undefined && 'right' in body ? body.right : result.body, validationErrors: body !== undefined && 'left' in body ? body.left : undefined, where: 'response:statuscode' } 
        }
        if (body === undefined || 'left' in body)
        {
            return {success: body === undefined, statusCode: result.statusCode.toString(), status, headers: result.headers, left: result.body, validationErrors: body?.left, where: 'response:body' }
        }
        return {success: true as const, statusCode: result.statusCode.toString(), status, headers: result.headers, right: result.body }
    }
}

",
    "examples/edges/client.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as GetEmployeesResponseDefaultValidator } from './schemas/get-employees-response-default.schema.js'
import { validate as GetEmployeesResponse200Validator } from './schemas/get-employees-response200.schema.js'

export interface GetEmployeesResponse200 {
    fullTime?: (boolean | undefined)
    id?: (number | undefined)
    name?: (string | undefined)
}

export const GetEmployeesResponse200 = {
    validate: GetEmployeesResponse200Validator as ValidateFunction<GetEmployeesResponse200>,
    get schema() { return GetEmployeesResponse200.validate.schema},
    get errors() { return GetEmployeesResponse200.validate.errors ?? undefined },
    is: (o: unknown): o is GetEmployeesResponse200 => GetEmployeesResponse200.validate(o) === true,
    parse: (o: unknown): { right: GetEmployeesResponse200 } | { left: DefinedError[] } => 
    {
        if(GetEmployeesResponse200.is(o)) {
            return { right: o }
        }
        return { left: (GetEmployeesResponse200.errors ?? []) as DefinedError[] }
    },
} as const


export interface GetEmployeesResponseDefault {
    fullTime?: (boolean | undefined)
    id?: (number | undefined)
    name?: (string | undefined)
}

export const GetEmployeesResponseDefault = {
    validate: GetEmployeesResponseDefaultValidator as ValidateFunction<GetEmployeesResponseDefault>,
    get schema() { return GetEmployeesResponseDefault.validate.schema},
    get errors() { return GetEmployeesResponseDefault.validate.errors ?? undefined },
    is: (o: unknown): o is GetEmployeesResponseDefault => GetEmployeesResponseDefault.validate(o) === true,
    parse: (o: unknown): { right: GetEmployeesResponseDefault } | { left: DefinedError[] } => 
    {
        if(GetEmployeesResponseDefault.is(o)) {
            return { right: o }
        }
        return { left: (GetEmployeesResponseDefault.errors ?? []) as DefinedError[] }
    },
} as const


",
    "examples/edges/schemas/get-employees-response-default.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"GetEmployeesResponseDefault","type":"object","properties":{"fullTime":{"type":"boolean"},"id":{"type":"integer"},"name":{"type":"string"}},"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){if(data.fullTime !== undefined){const _errs2 = errors;if(typeof data.fullTime !== "boolean"){validate10.errors = [{instancePath:instancePath+"/fullTime",schemaPath:"#/properties/fullTime/type",keyword:"type",params:{type: "boolean"},message:"must be boolean"}];return false;}var valid0 = _errs2 === errors;}else {var valid0 = true;}if(valid0){if(data.id !== undefined){let data1 = data.id;const _errs4 = errors;if(!(((typeof data1 == "number") && (!(data1 % 1) && !isNaN(data1))) && (isFinite(data1)))){validate10.errors = [{instancePath:instancePath+"/id",schemaPath:"#/properties/id/type",keyword:"type",params:{type: "integer"},message:"must be integer"}];return false;}var valid0 = _errs4 === errors;}else {var valid0 = true;}if(valid0){if(data.name !== undefined){const _errs6 = errors;if(typeof data.name !== "string"){validate10.errors = [{instancePath:instancePath+"/name",schemaPath:"#/properties/name/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs6 === errors;}else {var valid0 = true;}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/edges/schemas/get-employees-response200.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"GetEmployeesResponse200","type":"object","properties":{"fullTime":{"type":"boolean"},"id":{"type":"integer"},"name":{"type":"string"}},"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){if(data.fullTime !== undefined){const _errs2 = errors;if(typeof data.fullTime !== "boolean"){validate10.errors = [{instancePath:instancePath+"/fullTime",schemaPath:"#/properties/fullTime/type",keyword:"type",params:{type: "boolean"},message:"must be boolean"}];return false;}var valid0 = _errs2 === errors;}else {var valid0 = true;}if(valid0){if(data.id !== undefined){let data1 = data.id;const _errs4 = errors;if(!(((typeof data1 == "number") && (!(data1 % 1) && !isNaN(data1))) && (isFinite(data1)))){validate10.errors = [{instancePath:instancePath+"/id",schemaPath:"#/properties/id/type",keyword:"type",params:{type: "integer"},message:"must be integer"}];return false;}var valid0 = _errs4 === errors;}else {var valid0 = true;}if(valid0){if(data.name !== undefined){const _errs6 = errors;if(typeof data.name !== "string"){validate10.errors = [{instancePath:instancePath+"/name",schemaPath:"#/properties/name/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs6 === errors;}else {var valid0 = true;}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
  },
}
`;

exports[`locals 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/edges/locals.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

type Days2 = 'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday' | 'sunday'

export interface Days {
    days: (Days2)[]
}

type Local2 = string

type Local = string

export type Ref1 = Local2

export type Ref2 = Local

export type Unevaluated = string

",
  },
}
`;

exports[`optional 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/edges/optional.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

export type MadeOptionalRef = (Original | undefined)

export type OptionalString = (string | undefined)

export type OptionalUnknown = unknown

export type Original = string

",
  },
}
`;

exports[`validator 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/edges/schemas/validator.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Validator","type":"object","properties":{"foo":{"type":"string"}},"required":["foo"],"additionalProperties":true};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((data.foo === undefined) && (missing0 = "foo")){validate10.errors = [{instancePath,schemaPath:"#/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {if(data.foo !== undefined){if(typeof data.foo !== "string"){validate10.errors = [{instancePath:instancePath+"/foo",schemaPath:"#/properties/foo/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/edges/validator.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as ValidatorValidator } from './schemas/validator.schema.js'

export interface Validator {
    foo: string
}

export const Validator = {
    validate: ValidatorValidator as ValidateFunction<Validator>,
    get schema() { return Validator.validate.schema},
    get errors() { return Validator.validate.errors ?? undefined },
    is: (o: unknown): o is Validator => Validator.validate(o) === true,
    parse: (o: unknown): { right: Validator } | { left: DefinedError[] } => 
    {
        if(Validator.is(o)) {
            return { right: o }
        }
        return { left: (Validator.errors ?? []) as DefinedError[] }
    },
} as const


",
  },
}
`;
